File: web/admin/src/types/feature-request.ts
/**
 * 要望リクエスト機能 - ローカル型定義
 *
 * NOTE: バックエンド API 実装後は types/api/v1/ 配下の自動生成型に移行する。
 */

export const FeatureRequestStatus = {
  /** 受付中 */
  Waiting: 1,
  /** 検討中 */
  Reviewing: 2,
  /** 採用決定 */
  Adopted: 3,
  /** 開発中 */
  InProgress: 4,
  /** 完了 */
  Done: 5,
  /** 却下 */
  Rejected: 6,
} as const
export type FeatureRequestStatus = typeof FeatureRequestStatus[keyof typeof FeatureRequestStatus]

export const FeatureRequestCategory = {
  /** UI/UX改善 */
  UI: 1,
  /** 新機能 */
  Feature: 2,
  /** パフォーマンス改善 */
  Performance: 3,
  /** その他 */
  Other: 4,
} as const
export type FeatureRequestCategory = typeof FeatureRequestCategory[keyof typeof FeatureRequestCategory]

export const FeatureRequestPriority = {
  /** 低 */
  Low: 1,
  /** 中 */
  Medium: 2,
  /** 高 */
  High: 3,
} as const
export type FeatureRequestPriority = typeof FeatureRequestPriority[keyof typeof FeatureRequestPriority]

export interface FeatureRequest {
  id: string
  title: string
  description: string
  category: FeatureRequestCategory
  priority: FeatureRequestPriority
  status: FeatureRequestStatus
  /** 管理者コメント */
  note: string
  submittedBy: string
  submitterName: string
  createdAt: number
  updatedAt: number
}

export interface CreateFeatureRequestInput {
  title: string
  description: string
  category: FeatureRequestCategory
  priority: FeatureRequestPriority
  submittedBy: string
  submitterName: string
}

export interface UpdateFeatureRequestInput {
  status: FeatureRequestStatus
  note: string
}

File: api/internal/gateway/admin/v1/service/feature_request.go
package service

import (
	"github.com/and-period/furumaru/api/internal/gateway/admin/v1/types"
	"github.com/and-period/furumaru/api/internal/messenger/entity"
	"github.com/and-period/furumaru/api/pkg/jst"
)

type FeatureRequest struct {
	types.FeatureRequest
}

type FeatureRequests []*FeatureRequest

func NewFeatureRequest(featureRequest *entity.FeatureRequest) *FeatureRequest {
	return &FeatureRequest{
		FeatureRequest: types.FeatureRequest{
			ID:            featureRequest.ID,
			Title:         featureRequest.Title,
			Description:   featureRequest.Description,
			Category:      types.FeatureRequestCategory(featureRequest.Category),
			Priority:      types.FeatureRequestPriority(featureRequest.Priority),
			Status:        types.FeatureRequestStatus(featureRequest.Status),
			Note:          featureRequest.Note,
			SubmittedBy:   featureRequest.SubmittedBy,
			SubmitterName: "",
			CreatedAt:     jst.Unix(featureRequest.CreatedAt),
			UpdatedAt:     jst.Unix(featureRequest.UpdatedAt),
		},
	}
}

func (f *FeatureRequest) Fill(submitterName string) {
	f.SubmitterName = submitterName
}

func (f *FeatureRequest) Response() *types.FeatureRequest {
	return &f.FeatureRequest
}

func NewFeatureRequests(featureRequests entity.FeatureRequests) FeatureRequests {
	res := make(FeatureRequests, len(featureRequests))
	for i := range featureRequests {
		res[i] = NewFeatureRequest(featureRequests[i])
	}
	return res
}

func (fs FeatureRequests) Fill(admins map[string]*Admin) {
	for _, f := range fs {
		if admin, ok := admins[f.SubmittedBy]; ok {
			f.SubmitterName = admin.Name()
		}
	}
}

func (fs FeatureRequests) SubmittedByIDs() []string {
	ids := make([]string, 0, len(fs))
	seen := make(map[string]struct{}, len(fs))
	for _, f := range fs {
		if _, ok := seen[f.SubmittedBy]; !ok {
			seen[f.SubmittedBy] = struct{}{}
			ids = append(ids, f.SubmittedBy)
		}
	}
	return ids
}

func (fs FeatureRequests) Response() []*types.FeatureRequest {
	res := make([]*types.FeatureRequest, len(fs))
	for i := range fs {
		res[i] = fs[i].Response()
	}
	return res
}
